### DISCUSSION TITLE
A global error definition to an api (to a backend as whole, not individual endpoint)

### DISCUSSION THREAD

**Discussion by QzCurious** [ecyrbe/zodios#203] [Q&A] [answered]
I'm wondering if it's a good idea to have some global, or a fallback level, error for some general purpose error like 401, 403.
Currently, I'm not able to implement it with zodios. If anyone can share how you handle with it?

For example, 401 is happened when session token is expired or user is logged out for whatever reason. I don't want copy and paste an error instance to each endpoint.

OR, if there you have a good reason I should do the copy, please share me why.

**Comment by ecyrbe** [comment]
This is how i do it :
```ts
const commonErrors = makeErrors([{
  status : 401,
  schema: z.object( ... )
},{
  status: 'default',
  schema: z.object( ... )
}
]);

const api = makeApi([{
  mehod: 'get',
  path: '/users/:id',
  ...
  errors: [
    ...comonErrors,
    {
        status: 404,
        schema: ....
     }
  ]
},{
  mehod: 'get',
  path: '/users',
  ...
  errors: comonErrors
}]);

**Reply by QzCurious** [reply]
Yes, the `comonErrors` would appear in almost all endpoint. And it what I meant for copying (Sorry I didn't include code snippet this time). Thought it's like a one time job for each endpoint. ~~But I feel it too tedious.~~

**Comment by QzCurious** [comment, marked as answer]
Rethinked about it. From backend developer point of view. They also explict **guard** each endpoint that require user login.

Here is why I should do it too:

For setup part:

1. Not all endpoint require user login
2. The copying of `comonErrors` (or even copying 401 error and 403 error separately) for each endpoint is just a onetime job
3. Because we do it explicitly, it is more intuitive for us to know what an endpoint would throw
4. What we should is to encapsulate the logic, so the guard function, but not to over-engineer the architecture

For request and catching part:
The way [thelinuxlich do](https://github.com/ecyrbe/zodios/discussions/203#discussioncomment-3972985) is pretty consice.

**Comment by thelinuxlich** [comment]
I handle errors in Zodios like this:

```ts
const [err, response] = await goTry(zodiosAPI.getSomething())
throwIf(err)
// do something with response
```

**Reply by QzCurious** [reply]
I'm assuming that `goTry` is handling the global things right?

**Reply by thelinuxlich** [reply]
It's a library I developed to have Go-like error handling: https://github.com/thelinuxlich/go-go-try
